blockchain/blockchain.go

```diff
@@ -240,9 +254,228 @@ func NewBlockChain(db database.DBManager, cacheConfig *CacheConfig, chainConfig
+	database.DBManager
+}
+
+func (td *stateTrieMigrationDB) ReadCachedTrieNode(hash common.Hash) ([]byte, error) {
+	return td.ReadCachedTrieNodeFromNew(hash)
+}
+func (td *stateTrieMigrationDB) ReadCachedTrieNodePreimage(secureKey []byte) ([]byte, error) {
+	return td.ReadCachedTrieNodePreimageFromNew(secureKey)
+}
+
+func (td *stateTrieMigrationDB) ReadStateTrieNode(key []byte) ([]byte, error) {
+	return td.ReadStateTrieNodeFromNew(key)
+}
+
+func (td *stateTrieMigrationDB) HasStateTrieNode(key []byte) (bool, error) {
+	return td.HasStateTrieNodeFromNew(key)
+}
+
+func (td *stateTrieMigrationDB) ReadPreimage(hash common.Hash) []byte {
+	return td.ReadPreimageFromNew(hash)
+}
+
+func (bc *BlockChain) migrateState(rootHash common.Hash) error {
+	srcCachedDB := bc.StateCache().TrieDB()
+	targetDB := statedb.NewDatabase(&stateTrieMigrationDB{bc.db})
+
+	trieSync := state.NewStateSync(rootHash, targetDB.DiskDB())
+	maxBatchSize := 10000
+	queue := append([]common.Hash{}, trieSync.Missing(maxBatchSize)...)
+
+	committedCnt := 0
+
+	for trieSync.Pending() > 0 {
+		// TODO-Klaytn refine the status parameter (progress percentage, etc)
+		bc.committedCnt, bc.pendingCnt = committedCnt, trieSync.Pending()
+		logger.Warn("State migration progress", "committedCnt", bc.committedCnt, "pendingCnt", bc.pendingCnt)
+
+		results := make([]statedb.SyncResult, len(queue))
+
+		for i, hash := range queue {
+			data, err := srcCachedDB.NodeFromOld(hash)
+			if err != nil {
+				return fmt.Errorf("failed to retrieve node data for %x: %v", hash, err)
+			}
+
+			results[i] = statedb.SyncResult{Hash: hash, Data: data}
+		}
+
+		committedCnt += len(results)
+
+		if _, index, err := trieSync.Process(results); err != nil {
+			return fmt.Errorf("failed to process result #%d: %v", index, err)
+		}
+
+		if index, err := trieSync.Commit(targetDB.DiskDB().GetStateTrieMigrationDB()); err != nil {
+			return fmt.Errorf("failed to commit data #%d: %v", index, err)
+		}
+
+		queue = append(queue[:0], trieSync.Missing(maxBatchSize)...)
+
+		select {
+		case <-bc.stopMigration:
+			// TODO-Klaytn Revert DB.
+			// - copied new DB data to old DB.
+			// - remove new DB
+			return errors.New("stop state migration")
+		case <-bc.quit:
+			return nil
+		default:
+		}
+	}
+	logger.Info("completed to copy state migration", "resultCnt", committedCnt)
+
+	// Preimage Copy
+	// TODO-Klaytn consider to copy preimage
+
+	// Cross check that the two tries are in sync
+	// TODO-Klaytn consider to check Trie contents optionally
+	// TODO-Klaytn consider to check storage trie also
+	dirty, err := bc.checkTrieContents(targetDB, srcCachedDB, rootHash)
+	if err != nil || len(dirty) > 0 {
+		logger.Error("copied stated is invalid", "err", err, "len(dirty)", len(dirty))
+		// TODO-Klaytn Remove new DB and log.Error
+		if err != nil {
+			return err
+		}
+
+		return errors.New("copied state is not same with origin")
+	}
+
+	// TODO-Klaytn Swap DB and Remove Old DB
+	bc.db.FinishStateMigration()
```

Comment:
This should be private function.

Reply:
`FinishStateMigration()` is the function of Database. So it should be public function. Right?

Reply:
You are right :) It is not possible to implement if it is a private one.